src/System.Private.CoreLib/shared/System/Memory.cs

```diff
@@ -327,37 +360,51 @@ public Span<T> Span
         /// </summary>
         public unsafe MemoryHandle Pin()
         {
-            if (_index < 0)
-            {
-                Debug.Assert(_object != null);
-                return ((MemoryManager<T>)_object).Pin((_index & RemoveFlagsBitMask));
-            }
-            else if (typeof(T) == typeof(char) && _object is string s)
-            {
-                // This case can only happen if a ReadOnlyMemory<char> was created around a string
-                // and then that was cast to a Memory<char> using unsafe / marshaling code.  This needs
-                // to work, however, so that code that uses a single Memory<char> field to store either
-                // a readable ReadOnlyMemory<char> or a writable Memory<char> can still be pinned and
-                // used for interop purposes.
-                GCHandle handle = GCHandle.Alloc(s, GCHandleType.Pinned);
-                void* pointer = Unsafe.Add<T>(Unsafe.AsPointer(ref s.GetRawStringData()), _index);
-                return new MemoryHandle(pointer, handle);
-            }
-            else if (_object is T[] array)
+            // Just like the Span property getter, we have special support for a mutable Memory<char>
+            // that wraps an immutable String instance. This might happen if a caller creates an
+            // immutable ROM<char> wrapping a String, then uses Unsafe.As to create a mutable M<char>.
+            // This needs to work, however, so that code that uses a single Memory<char> field to store either
+            // a readable ReadOnlyMemory<char> or a writable Memory<char> can still be pinned and
+            // used for interop purposes.
+
+            // It's possible that the below logic could result in an AV if the struct
+            // is torn. This is ok since the caller is expecting to use raw pointers,
+            // and we're not required to keep this as safe as the other Span-based APIs.
+
+            object tmpObject = _object;
+            if (tmpObject != null)
             {
-                // Array is already pre-pinned
-                if (_length < 0)
+                if (typeof(T) == typeof(char) && tmpObject is string s)
```

Comment:
In the Span getter, you switched the similar code to using GetType and an unsafe cast for getting the string.  Is there a reason this differs / can't do the same thing?  Should the other one be converted back to match this? etc.

Reply:
The `Unsafe.As` construction has better performance so seemed more appropriate for the `Span` property getter. I didn't make the same change here because the `GCHandle.Alloc` call dominates, and I preferred readability over an insignificant performance gain. But if there's benefit to having the two code paths unified we can definitely do so.

Reply:
Ok, thanks.